samples/hello-oboe/src/main/cpp/PlayAudioEngine.cpp

```diff
@@ -178,27 +187,18 @@ PlayAudioEngine::onAudioReady(oboe::AudioStream *audioStream, void *audioData, i
     Trace::beginSection("numFrames %d, Underruns %d, buffer size %d",
                         numFrames, underrunCountResult.value(), bufferSize);
 
+    bool is16BitFormat = (audioStream->getFormat() == oboe::AudioFormat::I16);
     int32_t channelCount = audioStream->getChannelCount();
 
-    // If the tone is on we need to use our synthesizer to render the audio data for the sine waves
-    if (audioStream->getFormat() == oboe::AudioFormat::Float) {
-        if (mIsToneOn) {
-            for (int i = 0; i < channelCount; ++i) {
-                mOscillators[i].render(static_cast<float *>(audioData) + i, channelCount, numFrames);
-            }
-        } else {
-            memset(static_cast<uint8_t *>(audioData), 0,
-                   sizeof(float) * channelCount * numFrames);
-        }
-    } else {
-        if (mIsToneOn) {
-            for (int i = 0; i < channelCount; ++i) {
-                mOscillators[i].render(static_cast<int16_t *>(audioData) + i, channelCount, numFrames);
-            }
-        } else {
-            memset(static_cast<uint8_t *>(audioData), 0,
-                   sizeof(int16_t) * channelCount * numFrames);
-        }
+    // If the stream is 16-bit render into a float buffer then convert that buffer to 16-bit ints
+    float *outputBuffer = (is16BitFormat) ? mConversionBuffer : static_cast<float *>(audioData);
+
+    mSoundGenerator->renderAudio(outputBuffer, numFrames);
```

Comment:
What if numFrames is greater than the size of the mConversionBuffer?
I guess that won't happen in a callback.

Reply:
That's why I used `stream->getBufferCapacityInFrames` to determine the size of `mConversionBuffer` - could `numFrames` ever be bigger than that?